.ci/Jenkinsfile

```diff
@@ -0,0 +1,206 @@
+            P_BUILD_MATRIX_GIT_COMMIT: 'main', // TODO RJPP_BRANCH
+            P_DOCKERFILE: 'Dockerfile',
+            P_BUILD_CONTEXT: './docker/pytorch',
+            P_GIT_REPO: gitUrl,
+            P_GIT_COMMIT: gitCommit,
+            P_CPU_LIMIT: '4',
+            P_MEM_LIMIT: '15Gi',
+            P_TIMEOUT: '3600',
+            P_KANIKO_PUSH_FINAL: gitBranch == "main", // only push if we're on the main branch
+        ) { stagingImage -> dockerImagePostBuild(stagingImage) }
+    }
+    if (gitBranch != "main" && gitBranch != "dev") {
+        // if not on main or dev, run the pytest again.
+        jobs << [
+            'Lint': { -> 
+                runWithChecks(
+                    name: 'lint',
+                    title: 'Lint',
+                    summary: 'Static Analysis Checks',
+                ) {
+                    builds << build(
+                        job: 'scratch/lint',
+                        parameters: [
+                            string(name: 'P_CLOUD', value: pCloud),
+                            string(name: 'P_GIT_REPO', value: gitUrl),
+                            string(name: 'P_GIT_COMMIT', value: gitCommit),
+                            string(name: 'P_DOCKER_IMAGE', value: "mosaicml/pytorch:1.10.0_cpu-python3.9-ubuntu20.04"),
+                            string(name: 'P_TIMEOUT', value: '1800'),
+                            string(name: 'P_CPU_LIMIT', value: "2"),
+                            string(name: 'P_MEM_LIMIT', value: "7Gi"),
+                            string(name: 'PIP_EXTRA_DEPS', value: 'all'),
+                        ]
+                    )
+                }
+            },
+            'Python 3.7 - All': { -> runPytest("3.7", "all") },
+            'Python 3.8 - All': { -> runPytest("3.8", "all") },
+            'Python 3.9 - All': { -> runPytest("3.9", "all") },
+            'Python 3.9 - Dev': { -> runPytest("3.9", "dev") },
+            'Python 3.9 - All (GPU)': { -> runPytest("3.9", "all", true, "gpu") },
+            'Python 3.9 - All (GPU-Deepspeed)': { -> runPytest("3.9", "all", true, "deepspeed") },
+        ]        
+    }
+    try {
+        parallel(jobs)
+    }
+    finally {
+        stage('Merge Artifcats') {
+            node (pCloud) {
+                checkout scm  // checking out the SCM so the coverage report can load the source
+                builds.each { item -> 
+                    copyArtifacts(
+                        projectName: item.fullProjectName,
+                        selector: specific("${item.number}"),
+                        fingerprintArtifacts: true,
+                        optional: true,
+                    )
+                }
+
+                sh 'ls -al'
+
+                sh 'mkdir -p build/output/'
+
+                sh 'ls -al build/output/'
+                archiveArtifacts(artifacts: "build/output/*.xml", fingerprint: true, allowEmptyArchive: true)
```

Comment:
This is fine, for now, but Jenkins will not serve as our long term artifact repo - and there's no guarantee these will be retained

We need to either push things we want to keep to GH, S3, or we look at setting something like artifactory up

Reply:
Right now the only artifacts are coverage reports -- which I'm ok with losing. But yes, once we start doing real test regressions, we'll need to set up a real artifact repo.